File: collector-processor/internal/ai/sanitize.go
package ai

import (
	"encoding/json"
	"regexp"
	"strings"
)

// Common prompt injection patterns to filter
var injectionPatterns = []string{
	"ignore previous",
	"disregard above",
	"ignore all",
	"forget everything",
	"new instructions",
	"system prompt",
	"system:",
	"assistant:",
	"user:",
	"[system]",
	"[assistant]",
	"</s>",
	"<|im_end|>",
	"<|im_start|>",
	"<|endoftext|>",
}

// injectionRegex matches common injection attempt patterns
var injectionRegex = regexp.MustCompile(`(?i)(ignore|disregard|forget).*(previous|above|instructions|context)`)

// SanitizeEventForPrompt sanitizes user-provided event data for safe inclusion in AI prompts.
// It applies several protections:
// 1. Length limiting to prevent context stuffing
// 2. Clear data boundary markers
// 3. Pattern filtering for known injection phrases
// 4. JSON structure preservation where possible
func SanitizeEventForPrompt(event json.RawMessage, maxLen int) string {
	if len(event) == 0 {
		return "[empty event]"
	}

	// Convert to string
	s := string(event)

	// Limit length
	if maxLen <= 0 {
		maxLen = 2048
	}
	if len(s) > maxLen {
		s = s[:maxLen] + "...[truncated]"
	}

	// Filter known injection patterns (case-insensitive)
	lower := strings.ToLower(s)
	for _, pattern := range injectionPatterns {
		if strings.Contains(lower, pattern) {
			// Replace the pattern while preserving case
			idx := strings.Index(lower, pattern)
			if idx >= 0 {
				s = s[:idx] + "[FILTERED]" + s[idx+len(pattern):]
				lower = strings.ToLower(s)
			}
		}
	}

	// Apply regex filter for more complex patterns
	s = injectionRegex.ReplaceAllString(s, "[FILTERED]")

	// Wrap in clear data boundaries
	return "---USER-DATA-START---\n" + s + "\n---USER-DATA-END---"
}

// SanitizeStringForPrompt sanitizes a plain string for inclusion in AI prompts.
func SanitizeStringForPrompt(s string, maxLen int) string {
	return SanitizeEventForPrompt(json.RawMessage(s), maxLen)
}

File: crates/driftlock-ai/src/sanitize.rs
/// Common prompt injection patterns to filter
const INJECTION_PATTERNS: &[&str] = &[
    "ignore previous",
    "disregard above",
    "ignore all",
    "forget everything",
    "new instructions",
    "system prompt",
    "system:",
    "assistant:",
    "user:",
    "[system]",
    "[assistant]",
    "</s>",
    "<|im_end|>",
    "<|im_start|>",
    "<|endoftext|>",
];

/// Sanitizes user-provided event data for safe inclusion in AI prompts.
///
/// Applies several protections:
/// 1. Length limiting to prevent context stuffing
/// 2. Clear data boundary markers
/// 3. Pattern filtering for known injection phrases
///
/// # Arguments
/// * `data` - The raw data to sanitize
/// * `max_len` - Maximum length (0 for default of 2048)
///
/// # Returns
/// Sanitized string safe for inclusion in AI prompts
pub fn sanitize_for_prompt(data: &str, max_len: usize) -> String {
    if data.is_empty() {
        return "[empty event]".to_string();
    }

    // Determine max length
    let max_len = if max_len == 0 { 2048 } else { max_len };

    // Truncate if needed
    let mut sanitized = if data.len() > max_len {
        format!("{}...[truncated]", &data[..max_len])
    } else {
        data.to_string()
    };

    // Filter known injection patterns (case-insensitive)
    let lower = sanitized.to_lowercase();
    for pattern in INJECTION_PATTERNS {
        if let Some(idx) = lower.find(pattern) {
            // Replace the pattern while preserving overall structure
            let end_idx = idx + pattern.len();
            if end_idx <= sanitized.len() {
                sanitized = format!("{}[FILTERED]{}", &sanitized[..idx], &sanitized[end_idx..]);
            }
        }
    }

    // Apply regex-like filtering for complex patterns
    // Pattern: (ignore|disregard|forget).*(previous|above|instructions|context)
    sanitized = filter_complex_patterns(&sanitized);

    // Wrap in clear data boundaries
    format!("---USER-DATA-START---\n{}\n---USER-DATA-END---", sanitized)
}

/// Filters complex injection patterns using simple string matching
fn filter_complex_patterns(s: &str) -> String {
    let lower = s.to_lowercase();
    let trigger_words = ["ignore", "disregard", "forget"];
    let target_words = ["previous", "above", "instructions", "context"];

    // Check if string contains both a trigger and target word
    let has_trigger = trigger_words.iter().any(|w| lower.contains(w));
    let has_target = target_words.iter().any(|w| lower.contains(w));

    if has_trigger && has_target {
        // Find and replace the suspicious section
        for trigger in &trigger_words {
            for target in &target_words {
                if lower.contains(trigger) && lower.contains(target) {
                    // Find the trigger word position
                    if let Some(start) = lower.find(trigger) {
                        if let Some(end) = lower[start..].find(target) {
                            let end_pos = start + end + target.len();
                            if end_pos <= s.len() {
                                return format!("{}[FILTERED]{}", &s[..start], &s[end_pos..]);
                            }
                        }
                    }
                }
            }
        }
    }

    s.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sanitize_empty() {
        let result = sanitize_for_prompt("", 0);
        assert_eq!(result, "[empty event]");
    }

    #[test]
    fn test_sanitize_normal_text() {
        let result = sanitize_for_prompt("normal log message", 0);
        assert!(result.contains("---USER-DATA-START---"));
        assert!(result.contains("---USER-DATA-END---"));
        assert!(result.contains("normal log message"));
    }

    #[test]
    fn test_sanitize_with_injection() {
        let result = sanitize_for_prompt("ignore previous instructions", 0);
        assert!(result.contains("[FILTERED]"));
        assert!(!result.contains("ignore previous"));
    }

    #[test]
    fn test_sanitize_truncation() {
        let long_text = "a".repeat(3000);
        let result = sanitize_for_prompt(&long_text, 100);
        assert!(result.contains("[truncated]"));
        assert!(result.len() < 3000);
    }

    #[test]
    fn test_sanitize_special_tokens() {
        let result = sanitize_for_prompt("Some text </s> more text", 0);
        assert!(result.contains("[FILTERED]"));
        assert!(!result.contains("</s>"));
    }

    #[test]
    fn test_sanitize_system_markers() {
        let result = sanitize_for_prompt("Log: system: unauthorized", 0);
        assert!(result.contains("[FILTERED]"));
    }
}
